### DISCUSSION TITLE
Trade off between avoiding cycle detection and avoiding garbage

### DISCUSSION THREAD

**Discussion by tlively** [WebAssembly/stack-switching#25] [Ideas]
A few other folks and I spent some time refining our understanding of the design space here after the in-person CG meeting, so I wanted to share a summary of what we worked through. Additional ideas and observations welcome!

### Avoiding cycle detection in engines requires garbage in generator loops

To make it impossible to create unreachable cycles between suspended continuations, continuation references can be internally represented as pointers to single-use wrapper objects containing pointers to the underlying stack resources. The pointers in the wrapper objects are cleared whenever they are used to resume a continuation and a new wrapper object is allocated every time a continuation suspends. Setting up an unreachable cycle between continuations would involve resuming one of them to pass it a reference to the other, but that same resumption would break the cycle by clearing the reference that the other continuation holds. Allocating and freeing these wrapper objects may be a significant cost for uses such as tight generator loops.

### Garbage-free generator loops require cycle detection in engines

To avoid generating garbage in tight generator loops, the internal representation of continuation references must contain pointers to the underlying stack resources rather than to single-use wrapper objects. These pointers remaining valid across multiple resumes makes it possible to create unreachable cycles between resumable continuations that simple reference counting would be unable to free.

**Comment by RossTate** [comment]
I am not sure GC is a particularly informative consideration. For example, it is pretty easy to make all the fiber instructions operate on a table of fibers without `fiberref` being a first-class value. From what I've heard, this table approach lines up well with how non-WasmGC producers are likely to work with fibers. For example, GC languages compiling to linear memory via root scanning have indicated they need a table of all fibers so that they can scan them for refs-as-ints during their own implementation of GC. This also makes it easier for resource-concerned engines to cap how many fibers an application can have by limiting the maximum total size of fiber tables. For WasmGC producers/engines, there can be a small "combining" extension that makes `fiberref` a first-class value and adds the obvious counterparts, much like the function references proposal does now.

**Reply by dead-claudia** [reply]
> For example, GC languages compiling to linear memory via root scanning have indicated they need a table of all fibers so that they can scan them for refs-as-ints during their own implementation of GC.

Can you elaborate? Like what specifically within them is being scanned? I'm not quite sure what's being spelled out here.

**Reply by RossTate** [reply]
A number of GC languages have indicated they plan to target linear memory (implementing their own GC in linear memory) and would like WebAssembly/design#1459 (which has been approved for Phase 1) to facilitate finding `i32`s representing object references (presiding in linear memory). When combined with stack-switching, this requires being able to scan suspended stacks, as well as the active stack, and as such they need a table of all (suspended) stacks to enumerate over.

**Reply by dead-claudia** [reply]
Oh okay. Thanks for the explanation.

**Comment by dead-claudia** [comment]
I like the idea of making it a table. It's pretty easy for such non-GC producers to just track it with a free list. That does come with a caveat, though: [there's not really any way to shrink tables after](https://github.com/WebAssembly/memory-control/issues/3).

**Reply by RossTate** [reply]
Good point. In general, there seems to be a trend pushing for wasm to give more control to letting applications manage their own resources.

**Reply by dead-claudia** [reply]
@RossTate On an tangential note, it makes me wonder how useful the GC proposal is really, especially if/when refs get similar treatment to block memory. By that point, you could just use tables to wrap anything from the outside world.

I could still see it being useful in that GC'd languages wouldn't have to ship an entire collector as part of their result, but it wouldn't be necessary or even all that useful for most non-GC lanugages when compiled to WebAssembly.

**Comment by kayceesrk** [comment]
The proposal of using a wrapper to the underlying stack is exactly how OCaml implements suspended continuations, though the motivation for this choice is different. OCaml uses it to ensure "at most once" resumption of continuations. Allocation & GC are not the bottleneck in our use cases.

Does the table of fibers approach somehow avoid the problem of unreachable cycles between suspended continuations? I assume that continuations are merely indices into the table? If that's the case, it would not be possible to know whether there are cycles, would it? 

In addition, some other mechanism would be needed to ensure the "at most once"/one-shot resumption property for continuations and avoid the analogue of "use-after-free" bug. One approach to avoid the "use-after-free" bug might be to represent continuations as a pair of an index into the table and a counter, the counter is incremented every time the continuation is resumed. Before resuming a continuation, one would need to check that the counter value in the continuation reference matches the current counter value at that index in the table. With the wrapper-based approach, "user-after-free" bug is avoided at the cost of an allocation per suspension.

**Reply by fgmccabe** [reply]
The table of fibers is effectively delegating GC to the application. This is both good and bad.
Wrt cycles, there is a fairly strong case (at the moment) to be made for NOT garbage collecting unreachable computations: primarily, there may be other resources that are referenced in the computation that need to be explicitly released. Since it does not seem likely that WasmGC will have finalizers, this has to be managed by the application.

That 'counter' is exactly what is envisaged, I believe, in the 'fat pointer' approach.

**Reply by RossTate** [reply]
> Does the table of fibers approach somehow avoid the problem of unreachable cycles between suspended continuations? I assume that continuations are merely indices into the table? If that's the case, it would not be possible to know whether there are cycles, would it?

It does because no suspended fiber can have a direct reference to another fiber, only an index into a table. So if a fiber is removed from the table, then it is unreachable. In practice, this means it will be the application's responsibility to determine when a fiber in a table is no longer needed and therefore can be removed, and that seems to be what non-wasmgc applications want anyways.

> One approach to avoid the "use-after-free" bug might be to represent continuations as a pair of an index into the table and a counter, the counter is incremented every time the continuation is resumed. Before resuming a continuation, one would need to check that the counter value in the continuation reference matches the current counter value at that index in the table.

True, though one would need to use BigInt counters, or else there is a limit on switches from a given stack (and we have to formalize the semantics of the limit and agree on its cap).

**Reply by tlively** [reply]
> Wrt cycles, there is a fairly strong case (at the moment) to be made for NOT garbage collecting unreachable computations: primarily, there may be other resources that are referenced in the computation that need to be explicitly released. Since it does not seem likely that WasmGC will have finalizers, this has to be managed by the application.

If a fiber/continuation would be eligible to be garbage collected, then the application has no way to access it, so it should have already arranged for any resources it holds to be freed. Since garbage collection is not semantically observable, I don't see any downside to allowing unreachable cycles of first-class fibers/continuations to be collected.

**Reply by kayceesrk** [reply]
>  there is a fairly strong case (at the moment) to be made for NOT garbage collecting unreachable computations

If GCing unreachable cycles of suspended computations is not a concern, is the purpose of table of fibers only to enforce the one-shot property? If so, would it be possible to avoid the table completely? The suspended continuation references can directly refer to the stack. The engines can store the counter directly in the meta data of the stack object. Perhaps the table has other benefits that I am missing.

**Reply by tlively** [reply]
The table of fibers is analogous to the indirect function call table for producers that use Wasm linear memory. These producers are not set up to handle first-class reference types well and do not depend on any engine-provided GC because they implement languages with manual memory management or because they implement their own GC in linear memory. These producers will be happy to use fiber/continuation references indirectly without ever moving them out of the table.

Similarly, engines that do not want to provide WasmGC or implement any cycle-detecting GC algorithm can still provide continuations/fibers via indirect continuation tables just like they provide indirect function calls through tables but not through first-class function references. Since the references are always rooted in tables, memory management becomes a userspace problem.

**Reply by titzer** [reply]
The tradeoff is also analogous; `call_indirect` has an associated cast and doesn't serve the use case of programs that do assume GC well. In the GC proposal, we add `call_ref` to obviate the cast and to pave the way for GC'd closures in the future.

Yet to be clear, the function table MVP design was driven primarily by the lack of any opaque reference type in Wasm MVP. That, and to give functions an "address" as an affordance to C toolchains needing that. Now we do have a standard opaque reference type, and we are adding a new capability. Making it `int`ly typed from the beginning has to be better motivated than "C would *prefer* doing it this way". Typed tables are already a thing now; they need no specific rope to use indexes into tables on top of a typed reference.

Also, as Wasm is today, one can create cycles between instances because tables of functions are mutable. So there *has to be* a backup cycle collector for reference-counting instances. This is therefore not a difference in kind, but degree.

**Reply by rossberg** [reply]
Of course, there is an important difference between lifetime management of functions vs stacks/continuations: functions are essentially _static_ resources. You cannot create new ones after linking – except in dynamic linking scenarios, which a Wasm runtime may or may not support. Stacks on the other hand are allocated entirely _dynamically_.

Hence, it is not at all clear whether the table approach is equally adequate. Most applications will not care when static resources like functions or tables are collected, if at all, but reliable memory management will be quite relevant for dynamic resources like stacks.

Also, I don't follow what problem it actually solves. Using tables essentially defines the cycles away, by breaking them with an integer index. But that does not magically make the problem go away. In practice, a stack will have functions on it that close over these tables, so the cycles haven't disappeared, they just became more indirect. Clearly, that doesn't make it easier for an engine that cares about collecting them.

So tables are a red herring in this discussion. The question we need to answer is whether we care about enabling engines to recover abandoned stacks, e.g., in the face of erroneous programs. If the answer is yes, then tables would only get in the way. If the answer is no, then we don't need to take round-about ways like tables, and can as well make stacks leak directly.

**Reply by RossTate** [reply]
When working through how linear-memory-wasm programs would use stack switching (in any of the designs), it seems inevitable that they will use tables. They do not have a means to store a reference to a fiber/continuation/handler/whatever in linear memory, so they will store an index into some table. Thus it seems we have to make the table approach work at the least (though that's not to say other approaches can't work as well). Due to the issues you raise, the application will be responsible for explicitly clearing (i.e. setting to null, at the least) table entries in order to make the fiber/continuation/handler/whatever unreachable. The problem this approach solves is that it *forces* the application to do so (if it wants its resources cleaned up) *regardless* of whether the engine supports cycle detection or not. That is, it is a portable solution. Yes, it means a buggy application will have resource leaks, but it will *consistently* have the same leaks across engines, and it is not Wasm's responsibility to prevent applications being hurt by their own bugs.

**Reply by rossberg** [reply]
AFAICS, using a table of stacks/continuations already works fine, there is nothing special we'll need to do to support it. OTOH, there is no reason to require it.

> it means a buggy application will have resource leaks

Yes, exactly, and as I said above, the actual question we need to answer is whether that is acceptable. Everything else just follows. Unfortunately, I have heard rather opposite opinions on this, including at the face-to-face.

**Reply by RossTate** [reply]
> Unfortunately, I have heard rather opposite opinions on this, including at the face-to-face.

Oh, the concern I heard on this front was about bugs in applications not affecting them in some engines (e.g. ones with cycle detection) but then surprising them in other engines (e.g. ones with only reference counting). Could you repeat these opposing opinions?

**Reply by tlively** [reply]
> In practice, a stack will have functions on it that close over these tables, so the cycles haven't disappeared, they just became more indirect. Clearly, that doesn't make it easier for an engine that cares about collecting them.

The problem being solved by the continuation table idea is that some (but not all) engines don't care about or cannot collect cycles. This is the same set of engines that would choose not to support WasmGC because they don't want to or cannot implement a full GC. Such an engine could still support stack switching by allowing continuation references only inside of tables and providing instructions that operate on those continuation references via index. Leaky programs would run out of resources on these engines but not on engines that support first-class continuation references, but that seems fine.

**Reply by dead-claudia** [reply]
I do want to point out that there are algorithms for cycle collection in reference counting systems, like https://dl.acm.org/doi/10.5555/646158.680003 and https://dl.acm.org/doi/10.1145/1255450.1255453.

**Reply by rossberg** [reply]
@RossTate:
> Oh, the concern I heard on this front was about bugs in applications not affecting them in some engines

The main concern is that some systems want to be able to guard themselves against erroneous applications, which implies that they need to be able to reliably reclaim the app's memory after its termination, even if the application leaked. Otherwise, a buggy app could compromise the host, even if the host is otherwise able to abort and clear out misbehaved apps.

**Reply by rossberg** [reply]
@tlively:
> The problem being solved by the continuation table idea is that some (but not all) engines don't care about or cannot collect cycles.

The way this ultimately "solves" the problem is by saying that leaking stacks are not collected. If that's what we're after then we can state that just fine without involving tables (and it becomes no easier with them). I don't see what they add to the discussion. The problem isn't that stacks/continuations are first-class, the problem is that they are created dynamically, which is independent of how they are stored and referenced. (And of course, table entries are first-class anyway in the presence of table.get/set.)

**Reply by RossTate** [reply]
> The main concern is that some systems want to be able to guard themselves against erroneous applications, which implies that they need to be able to reliably reclaim the app's memory after its termination, even if the application leaked. Otherwise, a buggy app could compromise the host, even if the host is otherwise able to abort and clear out misbehaved apps.

Stack tables would seem to fall in the same category as reclaiming linear memory. If all applications referencing a stack table have terminated, then the stack table (and typically its contained stacks) can be reclaimed.

**Reply by tlively** [reply]
> (And of course, table entries are first-class anyway in the presence of table.get/set.)

I'm assuming that engines that would be interested in this approach wouldn't allow `table.get/set`, just like how MVP Wasm didn't have `table.get/set`.

**Reply by rossberg** [reply]
@tlively, I'm not sure what you envision, but table.get/set are standard now, so if we want to specify anything we need to specify it in a way that is meaningful in the context of the full language. (Also, telling engines to forego the standard to solve their problem is equivalent to admitting that the language does not address their use case.)

**Reply by fgmccabe** [reply]
@rossberg ... does not address their use case
Indeed. But, in fact, many language implementers do not see being able to use WasmGC in the foreseeable future (go-lang, Scheme, Swift, Erlang to name a small sample).
However, I do think that this is kind of orthogonal to table set/get.

**Reply by rossberg** [reply]
@fgmccabe, sure, I don't think WasmGC has anything to do with this question.

**Comment by RossTate** [comment]
Another idea came to mind. We already know that to support multithreaded work-stealing, we will (eventually) need a notion of "shareable" stacks that contain only "shareable" references. To be more precise, it's okay for an *active* stack to (temporarily) have non-shareable references, but they need to be guaranteed to be off the stack by the time any suspension occurs. There's various ways to achieve this (so I won't go into illustrating one), but it occurs to me that non-wasmgc engines could reuse a slight variant of the same device to require a suspended stack to have *no* references. That is, such an active stack could temporarily use references (e.g. fetching a fiberref from a table in order to switch to it) so long as they're not stored on the stack (e.g. not stored in a local) at any point a suspension could occur. This would prevent cycles while not requiring all stack instructions to operate on a table. It would also be helpful for even WasmGC engines, as it would allow them to skip scanning such stacks except during "deep" phases cleaning up instance objects and code objects.

**Reply by fgmccabe** [reply]
64 bits should be enough. 2^64-1 ~= 10^20 seconds at 1 switch/nano amounts to 584 years before a roll over.

**Reply by RossTate** [reply]
@fgmccabe Wrong thread (though my fault—I didn't realize Discussions supported threads when I first posted this, so I moved the relevant portions to the relevant subthreads after the fact).

**Comment by dead-claudia** [comment]
I feel the overhead concerns are a little bit overblown regarding GC-style cycle collection for stacks. Realistically speaking, you're more likely to see hundreds at most for embedded runtimes, and at that scale, a mark and sweep algorithm is actually realistic. If they pack the mark/sweep bits separately, even 1000 of those is literally just 250 bytes (1 bit in-use + 1 bit allocated) of overhead compared to 4000 bytes just for the saved stack pointers alone. I don't see this scale presenting very many perf problems personally.

For larger embedded systems, I expect there to be enough headroom that most runtimes would be willing to support WasmGC anyways, so I doubt it'd be too much of an issue, but even 10k (also an exceptionally high number) isn't *too* bad to enumerate IMHO.

**Reply by fgmccabe** [reply]
In our design discussions we are targeting for a potential number of suspended computations of approximately 1 million. This is based on reported use cases for languages like Kotlin.

**Reply by dead-claudia** [reply]
That's fair. I was just focusing on the embedded side. At the scale of a million, I was thinking it's more likely the runtime would just be offering WasmGC as well (and so they could fuse stack management into that).

Of course, I'm open to being proven wrong here. And of course, while Kotlin's a garbage collected language, Rust and C++ certainly aren't, and those might be a little more compelling in terms of not doing reference counting.

Alternatively, just specifying cycles that don't go through GC'd structs may/must leak is a valid solution. Might not be the prettiest, but the simple way may very well be the best way.
